infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -205,37 +227,43 @@ public void apply(List<LogReplicationEntry> messages) throws Exception {
     }
 
     /**
-     * Read from the shadow table and write to the real table
-     * @param uuid: the real table uuid
+     * Read from the shadow table and write to the original stream
+     *
+     * @param streamId regular stream id
+     * @param seqNum sequence number to apply
+     * @param snapshot base snapshot timestamp
      */
-    public long applyShadowStream(UUID uuid, Long seqNum, long snapshot) {
-        UUID shadowUUID = uuidMap.get(uuid);
+    private long applyShadowStream(UUID streamId, long seqNum, long snapshot) {
+        UUID shadowStreamId = uuidMap.get(streamId);
+
+        // We can safely ignore trims on the shadow stream, as our protocol guarantees
+        // that the checkpointer/trimmer will not run during a snapshot sync. Therefore,
+        // any observed TrimmedException belongs to different snapshot cycles, for which,
+        // old data is guaranteed to be applied.
         StreamOptions options = StreamOptions.builder()
-                .ignoreTrimmed(false)
+                .ignoreTrimmed(true)
```

Comment:
Is it necessary to set ignore the trim? How about there is a bug in the plugin?
If the plugin works correct, we should not get trimmed exception. If the plugin doesn't work properly, the working shadow stream can still be trimmed and we get a trim exception, right?

Reply:
If we don't ignore the trim, we'll get TrimmedExceptions on a new snapshot cycle. At the end of the first snapshot cycle we signal the plugin to resume checkpoint/trim, therefore, by the second cycle, the shadow stream has been trimmed one or several times.

Reply:
Can we still open a stream that has been trimmed but no checkpoint? If there is bug in the plug in, we may lose data.
Can we bookkeeping the log timestamp where the current snapshot start to write to the shadow stream, while start to apply, we open the the stream and see the start timestamp is not trimmed?

Reply:
mmm, yeah I think this is something we can do. Let me try it out.

Reply:
fyi if you plan to do bookkeeping note that the snashotSync plugin already captures this info. please feel free to add a getter method to get the wall clock timestamp when the last snapshot sync requested checkpoint freezing.

Reply:
I mean the log sequence number not the wall time. I am still curious if the stream can be opened if it has been trimmed without checkpointing.